application/src/main/java/org/thingsboard/server/actors/ruleChain/DefaultTbContext.java

```diff
@@ -172,8 +174,27 @@ public void tellSelf(TbMsg msg, long delayMs) {
 
     @Override
     public void input(TbMsg msg, RuleChainId ruleChainId) {
-        msg.pushToStack(nodeCtx.getSelf().getRuleChainId(), nodeCtx.getSelf().getId());
-        nodeCtx.getChainActor().tell(new RuleChainInputMsg(ruleChainId, msg));
+        if (!msg.isValid()) {
+            return;
+        }
+        TbMsg inputMsg = msg.copyWithRuleChainId(ruleChainId);
+        inputMsg.pushToStack(nodeCtx.getSelf().getRuleChainId(), nodeCtx.getSelf().getId());
+        TransportProtos.ToRuleEngineMsg toReMsg = TransportProtos.ToRuleEngineMsg.newBuilder()
```

Comment:
below part are very similar to enqueue. please, try to reuse an existing enqueue methods

Reply:
Yep, enqueue is similar but not suitable for this case, but ok, I will extract duplicated code.